File: apps/guzo-api/src/main/java/et/guzo/service/PickupCodeService.java
package et.guzo.service;

import et.guzo.common.ApiException;
import et.guzo.domain.entity.Package;
import et.guzo.repository.PackageRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;

@Service
@RequiredArgsConstructor
public class PickupCodeService {

    private static final SecureRandom RANDOM = new SecureRandom();
    private final PackageRepository packageRepository;

    @Transactional
    public Package assignCodes(Package pkg) {
        if (pkg.getPickupPin() == null || pkg.getPickupPin().isBlank()) {
            String pin;
            do {
                pin = String.format("%06d", RANDOM.nextInt(1_000_000));
            } while (packageRepository.findByPickupPin(pin).isPresent());
            pkg.setPickupPin(pin);
        }
        if (pkg.getQrCode() == null || pkg.getQrCode().isBlank()) {
            pkg.setQrCode("GUZO:" + pkg.getTrackingNumber() + ":" + pkg.getPickupPin());
        }
        return packageRepository.save(pkg);
    }

    @Transactional(readOnly = true)
    public Package verifyPickup(String trackingOrQr, String pin) {
        String resolvedPin = pin != null ? pin.trim() : null;
        if ((trackingOrQr == null || trackingOrQr.isBlank()) && resolvedPin != null && !resolvedPin.isBlank()) {
            return packageRepository.findByPickupPin(resolvedPin)
                .orElseThrow(() -> ApiException.notFound("Parcel not found"));
        }

        final String reference;
        final String pinForCheck;
        if (trackingOrQr != null && trackingOrQr.startsWith("GUZO:")) {
            String[] parts = trackingOrQr.split(":");
            if (parts.length >= 3) {
                reference = parts[1];
                pinForCheck = parts[2];
            } else {
                reference = trackingOrQr;
                pinForCheck = resolvedPin;
            }
        } else {
            reference = trackingOrQr;
            pinForCheck = resolvedPin;
        }
        Package pkg = packageRepository.findByTrackingNumber(reference)
            .or(() -> packageRepository.findByPickupPin(pinForCheck))
            .orElseThrow(() -> ApiException.notFound("Parcel not found"));
        if (pinForCheck != null && pkg.getPickupPin() != null && !pkg.getPickupPin().equals(pinForCheck)) {
            throw ApiException.badRequest("Invalid pickup PIN");
        }
        return pkg;
    }
}

File: apps/server/src/modules/branches/pickup-code.service.ts
import { prisma } from '@delivery/database';
import { ApiError } from '../../utils/ApiError.js';

function randomPin(): string {
  return String(Math.floor(Math.random() * 1_000_000)).padStart(6, '0');
}

export async function assignPickupCodes(packageId: string) {
  const existing = await prisma.package.findUnique({ where: { id: packageId } });
  if (!existing) throw ApiError.notFound('Package not found');

  let pin = randomPin();
  for (let i = 0; i < 10; i++) {
    const clash = await prisma.package.findFirst({ where: { pickupPin: pin, NOT: { id: packageId } } });
    if (!clash) break;
    pin = randomPin();
  }

  return prisma.package.update({
    where: { id: packageId },
    data: {
      pickupPin: pin,
      qrCode: `GUZO:${existing.trackingNumber}:${pin}`,
    },
  });
}

export async function verifyPickup(reference: string, pin?: string | null) {
  let tracking = reference.trim();
  let resolvedPin = pin?.trim() ?? null;

  if (!tracking && resolvedPin) {
    const byPin = await prisma.package.findFirst({ where: { pickupPin: resolvedPin } });
    if (!byPin) throw ApiError.notFound('Parcel not found');
    return byPin;
  }

  if (tracking.startsWith('GUZO:')) {
    const parts = tracking.split(':');
    if (parts.length >= 3) {
      tracking = parts[1]!;
      resolvedPin = parts[2]!;
    }
  }

  const pkg =
    (await prisma.package.findUnique({ where: { trackingNumber: tracking } })) ??
    (resolvedPin ? await prisma.package.findFirst({ where: { pickupPin: resolvedPin } }) : null);

  if (!pkg) throw ApiError.notFound('Parcel not found');
  if (resolvedPin && pkg.pickupPin && pkg.pickupPin !== resolvedPin) {
    throw ApiError.badRequest('Invalid pickup PIN');
  }
  return pkg;
}
